src/app/shared/table-of-contents/table-of-contents.ts

```diff
@@ -43,92 +39,84 @@ export class TableOfContents implements OnInit {
               @Inject(DOCUMENT) private _document: Document) {
 
-    this._router.events.takeUntil(this._destroyed).subscribe((event) => {
-      if (event instanceof NavigationEnd) {
+    // Create new links and save root url at the end of navigation
+    this._router.events
+      .filter(event => event instanceof NavigationEnd)
+      .takeUntil(this._destroyed)
+      .subscribe(event => {
         const rootUrl = _router.url.split('#')[0];
         if (rootUrl !== this._rootUrl) {
-          this.links = this.createLinks();
+          this._links = this.createLinks();
           this._rootUrl = rootUrl;
         }
-      }
-    });
-
-    this._route.fragment.takeUntil(this._destroyed).subscribe(fragment => {
-      this._urlFragment = fragment;
-
-      const target = document.getElementById(this._urlFragment);
-      if (target) {
-        target.scrollIntoView();
-      }
-    });
+      });
+
+    // Scroll to section when the fragment changes
+    this._route.fragment
+      .takeUntil(this._destroyed)
+      .subscribe(fragment => {
+        this._urlFragment = fragment;
+        this.scrollFragmentIntoView();
+      });
   }
 
-  ngOnInit(): void {
-    // On init, the sidenav content element doesn't yet exist, so it's not possible
-    // to subscribe to its scroll event until next tick (when it does exist).
-    Promise.resolve().then(() => {
-      this._scrollContainer = this.container ?
-        this._document.querySelectorAll(this.container)[0] : window;
-
-      Observable.fromEvent(this._scrollContainer, 'scroll')
-        .takeUntil(this._destroyed)
-        .debounceTime(10)
-        .subscribe(() => this.onScroll());
-    });
+  ngOnInit() {
+    // Update active link after scroll events
+    this._scrollDispatcher.scrolled()
+      .takeUntil(this._destroyed)
+      .subscribe(scrollable =>
+        this._ngZone.run(() => {
+          this.updateScrollContainer(scrollable);
+          this.setActiveLink();
+        }));
   }
 
   ngOnDestroy(): void {
     this._destroyed.next();
+    this._destroyed.complete();
   }
 
-  updateScrollPosition(): void {
-    this.links = this.createLinks();
-
-    const target = document.getElementById(this._urlFragment);
-    if (target) {
-      target.scrollIntoView();
-    }
+  contentReady(): void {
+    this._links = this.createLinks();
+    this.scrollFragmentIntoView();
   }
 
-  /** Gets the scroll offset of the scroll container */
-  private getScrollOffset(): number {
-    const {top} = this._element.nativeElement.getBoundingClientRect();
-    if (typeof this._scrollContainer.scrollTop !== 'undefined') {
-      return this._scrollContainer.scrollTop + top;
-    } else if (typeof this._scrollContainer.pageYOffset !== 'undefined') {
-      return this._scrollContainer.pageYOffset + top;
-    }
+  /** Find the target from the url fragment and scroll it into view. */
+  private scrollFragmentIntoView(): void {
+    this._ngZone.runTask(() => {
```

Comment:
Why does this need to enter the zone if it's only updating the scroll position?

Reply:
Oops that was leftover from debugging something else. Removed.